osu.Game.Rulesets.Catch/Objects/CatchHitObject.cs

```diff
@@ -15,19 +15,20 @@ public abstract class CatchHitObject : HitObject, IHasXPosition, IHasComboInform
     {
         public const float OBJECT_RADIUS = 64;
 
-        private float x;
+        public float X { get; set; }
 
-        public float X
-        {
-            get => x + XOffset;
-            set => x = value;
-        }
+        internal bool XOffsetReversed { get; set; }
 
         /// <summary>
         /// A random offset applied to <see cref="X"/>, set by the <see cref="CatchBeatmapProcessor"/>.
         /// </summary>
         internal float XOffset { get; set; }
 
+        /// <summary>
+        /// The effective X position during gameplay, applying with HardRock offset.
+        /// </summary>
+        public float GameplayX => XOffsetReversed ? X - XOffset : X + XOffset;
```

Comment:
Given the potential impact this can have and that osu!lazer is used for live diffcalc, I'd like this to have an HR diffcalc test.
See https://github.com/ppy/osu/wiki/Generating-osu!stable-conversion-mappings for steps how to do this (use the `diffcalc-test` beatmap already in the tests). Select the HR mod from stable and it'll generate the relevant mappings.

Reply:
Created the test a bit tricky:
Create test case with HR in master to get correct SR in master, then paste it in this branch.

Surprisingly, HR gets the correct SR, but HRMR doesn't get the same,
```
Expected: 4.4882724679198711d
But was:  4.4848455546932069d
```

Investigating.

Reply:
For mirror only:
```
Expected: 4.050601681491468d
But was:  4.0506016929676321d
```

Seems like difficulty calculator itself having precision issue on this.

Reply:
MR diffcalc doesn't really matter, only interested in HR so it doesn't break existing stuff.

Reply:
Negate the X position in difficulty calculator:
None fails, MR passes, meaning it's a precision issue
HR fails, but HRMR cannot pass, meaning something is wrong in the interaction between HR and MR.